### PR TITLE
Fix race in adding connections to connsByPeer

### PR DESCRIPTION
**Pull request by Kubuxu**
Side A could already be running Identify with side B before B had the
connection in the connsByPeer, this lead to Connectedness returning
NotConnected and Identify failing.

### CODE REVIEW

File: p2p/net/mock/mock_peernet.go [hunk 1]
```diff
@@ -160,11 +160,41 @@ func (pn *peernet) connect(p peer.ID) (*conn, error) {
 func (pn *peernet) openConn(r peer.ID, l *link) *conn {
 	lc, rc := l.newConnPair(pn)
 	log.Debugf("%s opening connection to %s", pn.LocalPeer(), lc.RemotePeer())
+	addConnPair(pn, rc.net, lc, rc)
```
Comment (magik6k) [inline p2p/net/mock/mock_peernet.go:163]: Shouldn't this also have a rc.net.addConn below?

Comment (Kubuxu) [reply p2p/net/mock/mock_peernet.go:163]: It happens in the other goroutine in `remoteOpenConn`.

**Review by magik6k — APPROVED** [review]

File: p2p/net/mock/mock_peernet.go [hunk 2]
```diff
@@ -160,11 +160,41 @@ func (pn *peernet) connect(p peer.ID) (*conn, error) {
...
 	go rc.net.remoteOpenedConn(rc)
 	pn.addConn(lc)
 	return lc
 }
 
+// addConnPair adds connection to both peernets at the same time
+// must be followerd by pn1.addConn(c1) and pn2.addConn(c2)
+func addConnPair(pn1, pn2 *peernet, c1, c2 *conn) {
+	pn1.Lock()
+	pn2.Lock()
```
Comment (Stebalien) [inline p2p/net/mock/mock_peernet.go:174]: We need to sort these or we'll deadlock. Same with the other locks we're taking.

File: p2p/net/mock/mock_peernet.go [hunk 3]
```diff
@@ -160,11 +160,41 @@ func (pn *peernet) connect(p peer.ID) (*conn, error) {
...
+		if !found {
+			pn.connsByLink[c.link] = map[*conn]struct{}{}
+		}
+		pn.connsByLink[c.link][c] = struct{}{}
+	}
+	add(pn1, c1)
+	add(pn2, c2)
+
+	c1.notifLk.Lock()
+	c2.notifLk.Lock()
```
Comment (Stebalien) [inline p2p/net/mock/mock_peernet.go:193]: Are you sure we can do this? What if we connect from a notification handler?

**Review by Stebalien — CHANGES_REQUESTED** [review]

File: p2p/net/mock/mock_peernet.go (hunk 3 above)
Comment (Stebalien) [reply p2p/net/mock/mock_peernet.go:193]: Discussed. We're fine here because we just created these connections.

**Review by Stebalien — APPROVED** [review]

**Pull request merged**